Fix directory separator in Windows and New features

Milly opened the pull request:
Fix:
- Fix directory separator in Windows.

Feature:
- Add `dirSeparator` option.
  - Related vim's option 'shellslash' and 'completeslash'.
- Add `escapeChars` option.
  - Default is escaping space (NOT any other spaces, like Tab or LF).
  - Can escape ` #%` for command-mode.
- Add `absolute` option.
  - Default is true. Returns absolute path as before.
  - If set false. Return relative path from the current directory.

Test:
- Add tests for internal functions.

Milly committed:
Fix directory separator in Windows

Milly committed:
Add `absolute` param

Milly committed:
Add `escapeChars` param

tani commented on denops/@ddc-sources/path.ts:
```diff
@@ -45,28 +48,60 @@ export class Source extends BaseSource<Params, UserData> {
+      return dirSeparator;
+    }
+    return await denops.eval(
+      "!exists('+completeslash') ? 'slash' :" +
+        "&completeslash !=# '' ? &completeslash :" +
+        "&shellslash ? 'slash' : 'backslash'",
```
Instead of `Deno.eval`, consider using [denops_std/variable](https://github.com/vim-denops/deno-denops-std/tree/main/denops_std/variable).

Milly replied:
Fix it.

tani commented on denops/@ddc-sources/path.ts:
```diff
@@ -45,28 +48,60 @@ export class Source extends BaseSource<Params, UserData> {
     };
   }
+
+  private async _getDirSeparator(
```
Please avoid using `_` to hide the function name because we have already declare `private` at the head of statement.

Milly replied:
Fix it.

tani commented on doc/ddc-path.txt:
```diff
@@ -60,6 +60,21 @@ cmd		(string[])
+	  slash		A forward slash is used.
+	  backslash	A backward slash is used.
+	  (empty)	Auto detect by:
+			- When |'completeslash'| is not empty, the value is
+			  used for this option.
+			- When |'shellslash'| is enabled, 'slash' is used.
+			- When |'noshellslash'|, 'backslash' is used.
```
I am not sure. Is it possible to jump to the section declared the tags with quotations `'`. I guess we can just use `|completeslash|` for example.

Milly replied:
I made a mistake. References to options should enclose in single quotes only. eg.:

    'shellslash' (correct)
    |'shellslash'| (incorrect)
    |shellslash| (incorrect)

And, string should enclose in double quotes, because it is confusion to option.

tani replied:
Noted. Thank you.

tani commented on denops/@ddc-sources/path.ts:
```diff
@@ -67,9 +68,12 @@ export class Source extends BaseSource<Params, UserData> {
+
+    // change relative path to absolute
+    if (sourceParams.absolute) {
+      words = words.map((word) => `${cwd}/${word}`);
```
I doubt to use `/` for the separator if we add the support Windows file system.
Consider to use `path.join()` in Deno.

Milly replied:
Vim in Windows, Both `\` and `/` are used for the directory separator.
Therefore, it is first normalized with `/` and then converted according to the parameter values.

Vim-world Path cannot be handled by Deno's standard library.

tani replied:
I see. Usage of the separator of the pathname in Windows is a bit tricky. Vim's path is different from Deno's path. How about URL api? It can deal with such a clunky path.

```
> new URL(`file://\\a/b.txt`)
URL {
  href: "file:///a/b.txt",
  origin: "null",
  protocol: "file:",
  username: "",
  password: "",
  host: "",
  hostname: "",
  port: "",
  pathname: "/a/b.txt",
  hash: "",
  search: ""
}
```

Milly replied:
How to use URL api?
I don't understand what you intend to do.

tani replied:
True, I was stupid. Please never mind.

tani commented on doc/ddc-path.txt:
```diff
@@ -54,6 +54,13 @@ EXAMPLES						*ddc-path-examples*
+						     *ddc-path-param-absolute*
+absolute	(boolean)
+	If it is true, returns absolute path.
+	Otherwise, returns relative path from the current directory.
```
This is the grammartical issue.
1. Add the subject phrase. What returns something?
2. Add the article for each noun phrase, _e.g._, 'the absolute path'

Milly replied:
Fix it.

tani replied:
Thanks!

tani commented on denops/@ddc-sources/path.ts:
```diff
@@ -19,33 +19,68 @@ type Params = {
       }
     }
-    return Promise.resolve(arg.context.input.length);
+    return context.input.length;
```
`context.input.length` is `number` value. However, we declare that the function returns a value as a `Promise<number>`. Consider wrapping `Promise.resolve`

Milly replied:
`getCompletePosition()` is changed to async function, so `return <number>` is correct.
```diff
-  override getCompletePosition(
+  override async getCompletePosition(
```

Milly replied:
I found another bug in `getCompletePosition()`, so fix it and correct this.

tani replied:
Noted. Thanks

tani replied:
Thanks.

tani commented on denops/@ddc-sources/path.ts:
```diff
@@ -110,3 +147,9 @@ export class Source extends BaseSource<Params, UserData> {
   }
 }
+
+export const _internals = {
```
I am against using `_`. Consider using `export function ...`

Milly replied:
They are exporting to test internal functions.
This technique is also used in testing Deno's standard library.

tani commented on denops/@ddc-sources/path.ts:
```diff
@@ -19,33 +19,68 @@ type Params = {
   return true;
 }
 
+function getWordFilter(escapeChars: string): WordFilter {
+  if (escapeChars.length === 0) {
+    return (word) => word;
+  }
+  const escapeRegex = new RegExp(
+    `[${escapeChars.replaceAll(/[-\\\]]/g, "\\$&")}]`,
+    "g",
+  );
+  return (word) => word.replaceAll(escapeRegex, "\\$&");
+}
```
Consider using `escape()` defined as  vim builtin function.

Milly replied:
`getWordFilter()` called many times.
Implemented in Deno to offload communication overhead of denops and processing in Vim.

tani replied:
Noted. Then, How about the following code?

```js
function getWordFilter(escapeChars: string): WordFilter {
  const regexpsrc = `[${escapeChars.replaceAll(/[-\\\]]/g, "\\$&")}]`
  const escapeRegex = new RegExp(regexpsrc, "g");
  return (word) => word.replaceAll(escapeRegex, "\\$&");
}
```

We know that the Deno/ V8 is **very fast** interpreter. The creation of regexp objects is not so matter for the performance, is it? Further, we can expect cashing the regexp objects by the optimization of V8. Thus, the closure is no Moreover, '/[]/' matches nothing. I pefer the simpler code. Shall we believe the V8 optimization?

Milly replied:
Can you change the code block to a suggestion?

Milly replied:
Suggestion merged!

tani reviewed (changes requested):
Thank you for the remarkable effort to work it in Windows. I am impressed by the large test for the validation. I give the several suggestions. Cheers,

tani commented:
I am not sure. Do we truely have no such library to validate pathname in Deno/ Node?

Milly committed:
Fix enclosing for options or strings

- Enclose the options in single quotes. Do not enclose in bar (|).
- Enclose the strings in double quotes. Because confusing with options.

Milly committed:
Implement `getDirSeparator()` by denops_std lib

Milly committed:
Use test_suite library for test

tani commented on denops/@ddc-sources/path.ts:
```diff
@@ -19,33 +19,68 @@ type Params = {
   return true;
 }
 
+function getWordFilter(escapeChars: string): WordFilter {
+  if (escapeChars.length === 0) {
+    return (word) => word;
+  }
+  const escapeRegex = new RegExp(
+    `[${escapeChars.replaceAll(/[-\\\]]/g, "\\$&")}]`,
+    "g",
+  );
+  return (word) => word.replaceAll(escapeRegex, "\\$&");
+}
```
```suggestion
function getWordFilter(escapeChars: string): WordFilter {
  const regexpsrc = `[${escapeChars.replaceAll(/[-\\\]]/g, "\\$&")}]`
  const escapeRegex = new RegExp(regexpsrc, "g");
  return (word) => word.replaceAll(escapeRegex, "\\$&");
}
```

tani replied:
I have sent the suggestion. Cheers,

Milly committed:
Fix and move out `getCompletePosition()` to module function

Milly committed:
Simplified `getWordFilter()`

Co-authored-by: TANIGUCHI Masaya <ta2gch@gmail.com>

Milly committed:
Fix document of `absolute` param

tani commented:
Now, everything is fine. Thank you so much.
